src/topology/algebra/with_zero_topology.lean

```diff
@@ -0,0 +1,284 @@
+description is what is actually useful.
+
+We prove this topology is ordered and regular (in addition to be compatible with the monoid
+structure).
+
+All this is useful to extend a valuation to a completion. This is an abstract version of how the
+absolute value (resp. `p`-adic absolute value) on `ℚ` is extended to `ℝ` (resp. `ℚₚ`).
+
+## Implementation notes
+
+This topology is not defined as an instance since it may not be the desired topology on
+a linearly ordered commutative group with zero. You can locally activate this topology using
+`local attribute [instance] linear_ordered_comm_group_with_zero.topological_space`
+All other instances will (`ordered_topology`, `regular_space`, `has_continuous_mul`) then follow.
+
+-/
+
+open_locale topological_space
+open topological_space filter set
+
+namespace linear_ordered_comm_group_with_zero
+
+variables (Γ₀ : Type*) [linear_ordered_comm_group_with_zero Γ₀]
+
+/--The neighbourhoods around γ ∈ Γ₀, used in the definition of the topology on Γ₀.
+These neighbourhoods are defined as follows:
+A set s is a neighbourhood of 0 if there is an invertible γ₀ ∈ Γ₀ such that {γ | γ < γ₀} ⊆ s.
+If γ ≠ 0, then every set that contains γ is a neighbourhood of γ. -/
+def nhds_fun : Γ₀ → filter Γ₀ :=
+  (λ x : Γ₀, if x = 0 then ⨅ (γ₀ : units Γ₀), principal {γ | γ < γ₀} else pure x)
+
+/--The topology on a linearly ordered commutative group with a zero element adjoined.
+A subset U is open if 0 ∉ U or if there is an invertible element γ₀ such that {γ | γ < γ₀} ⊆ U. -/
+protected def topological_space : topological_space Γ₀ :=
+topological_space.mk_of_nhds (nhds_fun Γ₀)
+
+local attribute [instance] linear_ordered_comm_group_with_zero.topological_space
+
+/--The neighbourhoods {γ | γ < γ₀} of 0 form a directed set indexed by the invertible elements γ₀.-/
+lemma directed_lt : directed (≥) (λ γ₀ : units Γ₀, principal {γ : Γ₀ | γ < γ₀}) :=
+begin
+  intros γ₁ γ₂,
+  use linear_order.min γ₁ γ₂ ; dsimp only,
+  split ; rw [ge_iff_le, principal_mono] ; intros x x_in,
+  { calc x < ↑(linear_order.min γ₁ γ₂) : x_in
+        ... ≤ γ₁ : min_le_left γ₁ γ₂ },
+  { calc x < ↑(linear_order.min γ₁ γ₂) : x_in
+        ... ≤ γ₂ : min_le_right γ₁ γ₂ }
+end
+
+-- We need two auxilliary lemmas to show that nhds_fun accurately describes the neighbourhoods
+-- coming from the topology (that is defined in terms of nhds_fun).
+
+/--At all points of a linearly ordered commutative group with a zero element adjoined,
+the pure filter is smaller than the filter given by nhds_fun.-/
+private lemma pure_le_nhds_fun : pure ≤ nhds_fun Γ₀ :=
+λ x, by { by_cases hx : x = 0; simp [hx, nhds_fun] }
+
+/--For every point Γ₀, and every “neighbourhood” s of it (described by nhds_fun), there is a
+smaller “neighbourhood” t ⊆ s, such that s is a “neighbourhood“ of all the points in t.-/
+private lemma nhds_fun_ok : ∀ (x : Γ₀) (s ∈ nhds_fun Γ₀ x),
+  (∃ t ∈ nhds_fun Γ₀ x, t ⊆ s ∧ ∀ y ∈ t, s ∈ nhds_fun Γ₀ y) :=
+begin
+  intros x U U_in,
```

Comment:
Is there a reason you use `s` in the lemma statement but `U` in the proof, instead of doing?
```suggestion
private lemma nhds_fun_ok (x : Γ₀) (s) (hs : s ∈ nhds_fun Γ₀ x) :
  ∃ t ∈ nhds_fun Γ₀ x, t ⊆ s ∧ ∀ y ∈ t, s ∈ nhds_fun Γ₀ y :=
begin
```

Reply:
I have no idea. I wrote this two years ago. I fixed it and removed some non-terminal simps. It looks like I wrote that in a hurry...